### PR TITLE
[bench] textbook dcg/ndcg scoring module

### PR DESCRIPTION
**Pull request by norrietaylor**
## Summary

Adds a new pure-stdlib scoring module for the LongMemEval bench (Wave 1 of the [bench plan](https://github.com/norrietaylor/distillery/issues/233)):

- `src/distillery/eval/scoring.py` — `dcg`, `ndcg`, `evaluate_retrieval` (~30 LOC, `math` + stdlib only).
- `tests/eval/test_scoring.py` — **12-row spec table** + 5 dcg + 3 boundary tests, all `@pytest.mark.unit`.

The 12-row table *is* the spec: empty rankings, perfect ranking, worst within-k, missed beyond-k, partial match, all-correct, oversized k, empty correct_ids, near miss, spread across k=10, two-of-three at k=5, and last-rank at k=10. Each NDCG expectation is hand-computed from the textbook formula.

## Why reimplemented (not lifted)

Two reasons the upstream plan rules out lifting:

1. **mempalace's bench is in active methodology dispute** ([mempalace/mempalace#875](https://github.com/mempalace/mempalace/issues/875) — top-k bypass artifact, hand-tuning to failing questions, cross-system "category error"). Lifting code from a contested benchmark inherits the contested framing.
2. **`pytrec_eval` breaks `uvx distillery-mcp` installs** on toolchain-less hosts because of its C extension. The math is small enough (~30 lines of textbook IR formulas, Manning et al. *Introduction to Information Retrieval* §8.4 / Wikipedia "Discounted cumulative gain") that a 12-row test table fully specifies behaviour.

Module docstring cites Manning et al. and Wikipedia. Does not cite mempalace. No NOTICE entry needed — code is reimplemented, not lifted.

## Test plan

- [x] `pytest tests/eval/test_scoring.py -v` — 20/20 passed (12 spec rows + 5 dcg + 3 boundary).
- [x] `mypy --strict src/distillery/eval/scoring.py` — clean.
- [x] `ruff check src/ tests/` — clean.
- [x] `ruff format --check src/ tests/` — clean.
- [x] Touches only `src/distillery/eval/scoring.py` and `tests/eval/test_scoring.py` (no other files).
- [x] No external imports beyond `math` and stdlib.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Release Notes

* **New Features**
  * Added evaluation metrics for assessing retrieval and ranking quality, including standard IR metrics for measuring result relevance.

* **Tests**
  * Added comprehensive test suite with 12 scenarios covering boundary cases and metric edge conditions.

### CODE REVIEW

**Comment by norrietaylor** [comment]
@coderabbitai review

File: src/distillery/eval/scoring.py [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+            relevances.append(0.0)
+
+    # Ideal ranking: as many 1.0s up front as there are correct docs (capped at k).
+    n_correct = min(len(correct_ids), k)
+    ideal = [1.0] * n_correct
+
+    idcg = dcg(ideal, k)
+    if idcg == 0.0:
+        return 0.0
+    return dcg(relevances, k) / idcg
```
[the remark this answers is not in the document]
Comment (norrietaylor) [reply src/distillery/eval/scoring.py]: Fixed in 3caa0ae — `ndcg()` now intersects `correct_ids` with `corpus_ids` before sizing IDCG, and a 13th spec row (`correct_id_outside_corpus_k1`) pins the behaviour. Pre-fix that row would have produced NDCG ≈ 0.6131; post-fix it produces 1.0 as expected.

File: tests/eval/test_scoring.py [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+    def test_ndcg_zero_k_is_zero(self) -> None:
+        assert ndcg([0, 1], {"a"}, ["a", "b"], 0) == 0.0
+
+    def test_ndcg_negative_index_treated_as_irrelevant(self) -> None:
+        """Out-of-range indices contribute 0.0 and do not raise."""
+        # Index -1 / 99 should be treated as "not in correct_ids" and
+        # contribute zero relevance, not an IndexError.
+        result = ndcg([99, 0], {"a"}, ["a"], 2)
+        # rel = [0.0, 1.0] → DCG = 0 + 1/log2(3); IDCG = 1/log2(2) = 1.0
+        assert result == pytest.approx(1.0 / math.log2(3))
```
[the remark this answers is not in the document]
Comment (norrietaylor) [reply tests/eval/test_scoring.py:248]: Fixed in 3caa0ae — `test_ndcg_negative_index_treated_as_irrelevant` now passes `[-1, 0]` (was `[99, 0]`) so the negative branch of the `0 <= idx < len(corpus_ids)` guard is actually exercised. The expected value is unchanged (`1.0 / log2(3)`).

**Comment by norrietaylor** [comment]
@coderabbitai resolve
